src/Compatibility/Microsoft.DotNet.PackageValidation/ApiCompatRunner.cs

```diff
@@ -0,0 +1,62 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.CodeAnalysis;
+using Microsoft.DotNet.ApiCompatibility;
+
+namespace Microsoft.DotNet.PackageValidation
+{
+    /// <summary>
+    /// Runs ApiCompat over different assembly tuples.
+    /// </summary>
+    public class ApiCompatRunner
+    {
+        private List<(Stream leftAssemblyStream, Stream rightAssemblyStream, string assemblyName, string compatibilityReason, string header)> queue = new();
+        private string _noWarn;
+        private (string, string)[] _ignoredDifferences;
+
+        public ApiCompatRunner(string noWarn, (string, string)[] ignoredDifferences)
+        {
+            _noWarn = noWarn;
+            _ignoredDifferences = ignoredDifferences;
+        }
+
+        /// <summary>
+        /// Runs the api compat for the tuples in the queue.
+        /// </summary>
+        /// <returns>The list api compat diagnostics.</returns>
+        public IEnumerable<ApiCompatDiagnostics> RunApiCompat()
+        {
+            List<ApiCompatDiagnostics> apiDifferences = new();
+            foreach (var apicompatTuples in queue)
+            {
+                // TODO: Add version check and proper way to display api compat error messages for each tuple.
+                // TODO: Add optimisations tuples.
+                // TODO: Run it Asynchronously.
+                IAssemblySymbol leftSymbols =  new AssemblySymbolLoader().LoadAssembly(apicompatTuples.assemblyName, apicompatTuples.leftAssemblyStream);
```

Comment:
Would it make sense to have an object that represents an assembly and knows how to load it?  Such an object could keep track of the name, it's source, other interesting metadata (like framework, runtime, etc) and could handle the different ways it might need to be loaded.

Reply:
Yeah, with my optimizations to compare multiple rights to the same left, I will introduce a new abstraction that will hold the symbol and some properties to describe it... we can expand it as we need it to include paths, tfm, etc.